From 83ad3261da595f02b4dadbc14d00223ffae0d72c Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Thu, 22 Oct 2026 16:40:02 +0200
Subject: [PATCH] multigpuswapchain: more correctly check the buffer type

The Nvidia workaround should only be applied when copying to a dmabuf in system
memory; copying from another GPU to the Nvidia GPU with Vulkan is perfectly ok.

This fixes EGL being used instead of Vulkan for Intel/AMD -> Nvidia copies.

BUG: 527591
BUG: 527604
CCBUG: 527544
(cherry picked from commit 1dd781dbc545b48604f118be623aa106530ac4d5)
---
 src/multigpuswapchain.cpp | 32 ++++++++++++++++++++------------
 1 file changed, 20 insertions(+), 12 deletions(-)

diff --git a/src/multigpuswapchain.cpp b/src/multigpuswapchain.cpp
index e8b8a96f009..8c9ed346f28 100644
--- a/src/multigpuswapchain.cpp
+++ b/src/multigpuswapchain.cpp
@@ -147,12 +147,18 @@ static std::optional<DrmFormat> chooseFormat(uint32_t inputFormat, const FormatM
 static const auto s_forceLinear = environmentVariableBoolValue("KWIN_VULKAN_FORCE_LINEAR_DST");
 static const auto s_allowVulkanCopy = environmentVariableBoolValue("KWIN_ALLOW_VULKAN_MGPU_COPIES");
 
+enum class BufferType {
+    DeviceMemory,
+    SystemMemory,
+};
+
 struct CopyRet
 {
     std::unique_ptr<MultiGpuCopy> copy;
     ModifierList sourceModifiers;
 };
 static std::optional<CopyRet> createCopy(RenderDevice *device,
+                                         BufferType targetType,
                                          uint32_t sourceFormat,
                                          const ModifierList &sourceModifiers,
                                          GraphicsBufferAllocator *allocator,
@@ -163,7 +169,7 @@ static std::optional<CopyRet> createCopy(RenderDevice *device,
     // use Vulkan if possible.
     // NOTE that using Vulkan to copy from an Nvidia GPU to a dmabuf in system memory
     // causes glitches due to a driver bug, so that path is disabled by default for now
-    if (device->vulkanDevice() && s_allowVulkanCopy.value_or(!device->isNvidia() || hostAllocator)) {
+    if (device->vulkanDevice() && s_allowVulkanCopy.value_or(!device->isNvidia() || (targetType == BufferType::SystemMemory && !hostAllocator))) {
         auto retModifiers = device->vulkanDevice()->transferFormats()[sourceFormat].intersected(sourceModifiers);
         if (!retModifiers.empty()) {
             const auto fmt = chooseFormat(sourceFormat, device->vulkanDevice()->transferFormats(), formats);
@@ -274,31 +280,33 @@ std::unique_ptr<MultiGpuSwapchain> MultiGpuSwapchain::createForSampling(RenderDe
             const auto alignment = std::lcm(*sourceDevice->vulkanDevice()->minImportedHostPointerAlignment(),
                                             *destination->vulkanDevice()->minImportedHostPointerAlignment());
             hostAllocator = std::make_shared<HostMemoryGraphicsBufferAllocator>(alignment);
-            firstCopy = createCopy(sourceDevice, format, {modifier}, hostAllocator.get(), destination->allImportableFormats(), options, hostAllocator);
+            firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, {modifier}, hostAllocator.get(),
+                                   destination->allImportableFormats(), options, hostAllocator);
             if (!firstCopy) {
                 hostAllocator.reset();
             }
         }
         if (!firstCopy && sysMemDevice) {
-            firstCopy = createCopy(sourceDevice, format, {modifier}, sysMemDevice->allocator(), destination->allImportableFormats(), options);
+            firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, {modifier},
+                                   sysMemDevice->allocator(), destination->allImportableFormats(), options);
         }
         if (!firstCopy) {
-            firstCopy = createCopy(sourceDevice, format, {modifier}, fallbackDevice->allocator(), destination->allImportableFormats(), options);
+            firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, {modifier}, fallbackDevice->allocator(), destination->allImportableFormats(), options);
         }
         if (!firstCopy) {
             return nullptr;
         }
-        auto secondCopy = createCopy(destination, firstCopy->copy->format(), {firstCopy->copy->modifier()}, destination->allocator(), importFormats, options, hostAllocator);
+        auto secondCopy = createCopy(destination, BufferType::DeviceMemory, firstCopy->copy->format(), {firstCopy->copy->modifier()}, destination->allocator(), importFormats, options, hostAllocator);
         if (!secondCopy) {
             return nullptr;
         }
         return std::make_unique<MultiGpuSwapchain>(std::move(firstCopy->copy), std::move(secondCopy->copy), format);
     } else {
         if (sysMemDevice) {
-            firstCopy = createCopy(sourceDevice, format, {modifier}, sysMemDevice->allocator(), importFormats, options);
+            firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, {modifier}, sysMemDevice->allocator(), importFormats, options);
         }
         if (!firstCopy) {
-            firstCopy = createCopy(sourceDevice, format, {modifier}, fallbackDevice->allocator(), importFormats, options);
+            firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, {modifier}, fallbackDevice->allocator(), importFormats, options);
         }
         if (!firstCopy) {
             return nullptr;
@@ -339,7 +347,7 @@ std::optional<MultiGpuSwapchain::AllocationInfo> MultiGpuSwapchain::createForSca
         if (!retModifiers.isEmpty()) {
             // We can use the source buffer directly, since it's already in system memory.
             options.scanout = true;
-            auto copy = createCopy(destination, format, modifiers, targetDevice->allocator(), importFormats, options);
+            auto copy = createCopy(destination, BufferType::SystemMemory, format, modifiers, targetDevice->allocator(), importFormats, options);
             if (copy) {
                 return AllocationInfo{
                     .swapchain = std::make_unique<MultiGpuSwapchain>(std::move(copy->copy), nullptr, format),
@@ -361,22 +369,22 @@ std::optional<MultiGpuSwapchain::AllocationInfo> MultiGpuSwapchain::createForSca
         const auto alignment = std::lcm(*sourceDevice->vulkanDevice()->minImportedHostPointerAlignment(),
                                         *destination->vulkanDevice()->minImportedHostPointerAlignment());
         hostAllocator = std::make_shared<HostMemoryGraphicsBufferAllocator>(alignment);
-        firstCopy = createCopy(sourceDevice, format, modifiers, hostAllocator.get(), destination->allImportableFormats(), options, hostAllocator);
+        firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, modifiers, hostAllocator.get(), destination->allImportableFormats(), options, hostAllocator);
         if (!firstCopy) {
             hostAllocator.reset();
         }
     }
     if (!firstCopy && sysMemDevice) {
-        firstCopy = createCopy(sourceDevice, format, modifiers, sysMemDevice->allocator(), destination->allImportableFormats(), options);
+        firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, modifiers, sysMemDevice->allocator(), destination->allImportableFormats(), options);
     }
     if (!firstCopy) {
-        firstCopy = createCopy(sourceDevice, format, modifiers, fallbackDevice->allocator(), destination->allImportableFormats(), options);
+        firstCopy = createCopy(sourceDevice, BufferType::SystemMemory, format, modifiers, fallbackDevice->allocator(), destination->allImportableFormats(), options);
     }
     if (!firstCopy) {
         return std::nullopt;
     }
     options.scanout = true;
-    auto secondCopy = createCopy(destination, format, {firstCopy->copy->modifier()}, targetDevice->allocator(), importFormats, options, hostAllocator);
+    auto secondCopy = createCopy(destination, BufferType::DeviceMemory, format, {firstCopy->copy->modifier()}, targetDevice->allocator(), importFormats, options, hostAllocator);
     if (!secondCopy) {
         return std::nullopt;
     }
-- 
GitLab

